aten/src/ATen/native/BinaryOps.cpp

```diff
@@ -881,11 +898,28 @@ Tensor minimum(const Tensor& self, const Tensor& other) {
 
 // binary min, alias for minimum
 Tensor& min_out(Tensor& result, const Tensor& self, const Tensor& other) {
-  return at::minimum_out(result, self, other);
+    return at::minimum_out(result, self, other);
 }
 
 Tensor min(const Tensor& self, const Tensor& other) {
-  return at::minimum(self, other);
+    return at::minimum(self, other);
+}
+
+Tensor& fmin_out(Tensor& result, const Tensor& self, const Tensor& other) {
+    auto self_isnan = self.isnan();
+    auto other_isnan = other.isnan();
+    auto both_nan = self_isnan.logical_and(other_isnan);
+    return at::minimum_out(result,
+                           at::where(self_isnan == both_nan, self, other.to(self.dtype())),
+                           at::where(other_isnan == both_nan, other, self.to(other.dtype())));
+}
+
+Tensor fmin(const Tensor& self, const Tensor& other) {
+    auto self_isnan = self.isnan();
+    auto other_isnan = other.isnan();
+    auto both_nan = self_isnan.logical_and(other_isnan);
+    return at::minimum(at::where(self_isnan == both_nan, self, other.to(self.dtype())),
```

Comment:
Similarly, this can be written more efficiently as: __other_isnan || (!self_isnan && self < other)__
```
auto mask = other.isnan().logical_or_(self.isnan().logical_not_().logical_and_(self < other));
return at::where(mask, self, other);
```
You'd have to do the type promotion beforehand which I think is preferable.

Reply:
In this case I think `(self >= other).logical_or(other.isnan())` would work?

However, I would like to recommend this be written like minimum and maximum to use TensorIterator. That will be much more performant and readable. It will require, however, that a gradient be added. The gradient for maximum and minimum should be a good guide. The dispatch will also have to change to no longer be a Math kernel.

The TensorIterator kernels can just be wrappers around `std::fmin` and `std::fmax` in C++. I propose we adopt Python's standard for floating-point NaN handling and treat all NaNs as identical, so in this case we don't need to worry about which NaN to return.

Reply:
Your equation works if we can assume that comparisons between NaN always return False. This is usually the case but I believe it is not guaranteed in the C++ standard. I agree that writing using TensorIterator will be better and more future proof.

Reply:
That's true, we should probably add a note that we assume IEEE 754.